org-code-javabuilder/theater/src/main/java/org/code/theater/GifWriter.java

```diff
@@ -0,0 +1,113 @@
+    }
+  }
+
+  /**
+   * Write the given image as the next frame of the gif.
+   *
+   * @param img BufferedImage
+   * @param delay delay in milliseconds until the next frame of the gif.
+   */
+  public void writeToGif(BufferedImage img, int delay) {
+    try {
+      this.writer.writeToSequence(
+          new IIOImage(img, null, getMetadataForFrame(delay, img.getType())), params);
+    } catch (IOException e) {
+      throw new InternalJavabuilderError(InternalErrorKey.INTERNAL_RUNTIME_EXCEPTION, e.getCause());
+    }
+  }
+
+  /**
+   * Close the gif stream and flush any remaining bytes. Any updates after close will throw an
+   * exception because the writer has been closed.
+   */
+  public void close() {
+    try {
+      this.writer.endWriteSequence();
+      this.imageOutputStream.flush();
+    } catch (IOException e) {
+      throw new InternalJavabuilderError(InternalErrorKey.INTERNAL_RUNTIME_EXCEPTION, e.getCause());
+    }
+  }
+
+  /**
+   * Update the metadata for the next frame with the given delay and image type. See Gif metadata
+   * specification here:
+   * https://javadoc.scijava.org/Java7/javax/imageio/metadata/doc-files/gif_metadata.html
+   *
+   * @param delay int delay in milliseconds after this frame
+   * @param imageType int
+   * @return Updated IIOMetadata for the next frame
+   */
+  private IIOMetadata getMetadataForFrame(int delay, int imageType) {
+    IIOMetadata metadata =
+        this.writer.getDefaultImageMetadata(
+            ImageTypeSpecifier.createFromBufferedImageType(imageType), params);
+    String metaFormatName = metadata.getNativeMetadataFormatName();
+    IIOMetadataNode root = (IIOMetadataNode) metadata.getAsTree(metaFormatName);
+
+    IIOMetadataNode graphicsControlExtensionNode = getNode(root, "GraphicControlExtension");
+    graphicsControlExtensionNode.setAttribute("disposalMethod", "none");
+    graphicsControlExtensionNode.setAttribute("userInputFlag", "FALSE");
+    graphicsControlExtensionNode.setAttribute("transparentColorFlag", "FALSE");
+    // delay is expected in hundredths of a second, divide milliseconds by 10
+    graphicsControlExtensionNode.setAttribute("delayTime", Integer.toString(delay / 10));
+    graphicsControlExtensionNode.setAttribute("transparentColorIndex", "0");
+
+    try {
+      metadata.setFromTree(metaFormatName, root);
+    } catch (IIOInvalidTreeException e) {
+      throw new InternalJavabuilderError(InternalErrorKey.INTERNAL_RUNTIME_EXCEPTION, e.getCause());
+    }
+    return metadata;
+  }
+
+  /**
+   * Find metadata node with the given name from the rootNode
+   *
+   * @param rootNode
+   * @param nodeName
+   * @return Requested IIOMetadataNode
+   */
+  private static IIOMetadataNode getNode(IIOMetadataNode rootNode, String nodeName) {
+    int nNodes = rootNode.getLength();
+    for (int i = 0; i < nNodes; i++) {
+      if (rootNode.item(i).getNodeName().equalsIgnoreCase(nodeName)) {
```

Comment:
Is there any possibility of name collisions here?

Reply:
I don't think so. The metadata is in a very specific format, so there should only be one node per name https://javadoc.scijava.org/Java7/javax/imageio/metadata/doc-files/gif_metadata.html